### PR TITLE
Emote cooldown framework (no more spam)

### PR DESCRIPTION
**Pull request by Hopekz**
Adds in a emote cooldown framework.
----
This means that emotes cannot be spammed anymore.

By default, a single emote cannot be repeated until 0.4 seconds have passed.
This an emote property and is able to be set per emote.
I left *flip and *spin spam able because it's sacred and must be preserved.

I used this new emote delay framework to set *deathgasp emote delay to 3.4 seconds.
Why i did this: I remember how angry AI's sometimes spam their death sound every 0.01 seconds on a macro. This change removes this. 

yes, I cannibalized parts of TG code for this PR and edited it to fit in our codebase.

**EDIT: Nich found what PR this was from:
This was cannibalized from tgstation/tgstation#47868**

#### Changelog

:cl:  Ninjanomnom , Hopek
rscadd: Emote cooldown framework has been added. This means you cannot spam a single emote in rapid succession (sorry macro kiddies)!
rscadd: Used this new emote cooldown framework to add a longer delay to *deathgasp . Take that spamming silicons!
/:cl:

### CODE REVIEW

File: code/modules/mob/emote.dm [hunk 1]
```diff
@@ -13,19 +13,27 @@
...
+	return FALSE
+
 
 /datum/emote/flip
 	key = "flip"
 	key_third_person = "flips"
 	restraint_check = TRUE
 	mob_type_allowed_typecache = list(/mob/living, /mob/dead/observer)
 	mob_type_ignore_stat_typecache = list(/mob/dead/observer)
+	cooldown = (0 SECONDS)
```
Comment (alexkar598) [inline code/modules/mob/emote.dm]: those () are not needed, macros are parsed before compile aka first

File: code/modules/mob/emote.dm [hunk 2]
```diff
@@ -38,6 +46,7 @@
...
```
Comment (alexkar598) [inline code/modules/mob/emote.dm]: those () are not needed, macros are parsed before compile aka first

File: code/modules/mob/living/emote.dm [hunk 3]
```diff
@@ -80,6 +80,7 @@
 	message_monkey = "lets out a faint chimper as it collapses and stops moving..."
 	message_simple =  "stops moving..."
 	stat_allowed = UNCONSCIOUS
+	cooldown = (3.4 SECONDS)
```
Comment (alexkar598) [inline code/modules/mob/living/emote.dm]: those () are not needed, macros are parsed before compile aka first

File: code/datums/emotes.dm [hunk 4]
```diff
@@ -75,6 +76,17 @@
 	else
 		user.visible_message(msg)
 
+/// For handling emote cooldown, return true to allow the emote to happen
+/datum/emote/proc/check_cooldown(mob/user, intentional)
```
Comment (alexkar598) [inline code/datums/emotes.dm]:
```suggestion
/datum/emote/proc/check_cooldown(mob/user, intentional, update=TRUE) 
```

File: code/datums/emotes.dm [hunk 5]
```diff
@@ -75,6 +76,17 @@
 	else
 		user.visible_message(msg)
 
+/// For handling emote cooldown, return true to allow the emote to happen
+/datum/emote/proc/check_cooldown(mob/user, intentional) 
+	if(!intentional)
+		return TRUE
+	if(user.emotes_used && user.emotes_used[src] + cooldown > world.time)
+		return FALSE
+	if(!user.emotes_used)
```
Comment (alexkar598) [inline code/datums/emotes.dm:87]:
```suggestion
	if(!update)
		return TRUE
	if(!user.emotes_used)
```

**Review by alexkar598 — CHANGES_REQUESTED** [review]

**Comment by Hopekz** [comment]
Testing Alex's changes before I commit.

**Comment by Hopekz** [comment]
Committed changes.

File: code/__DEFINES/components.dm [hunk 6]
```diff
@@ -166,6 +166,7 @@
 	#define SPEECH_LANGUAGE 5
 	#define SPEECH_IGNORE_SPAM 6
 	#define SPEECH_FORCED 7 */
+#define COMSIG_MOB_EMOTE "mob_emote" // from /mob/living/emote(): ()
```
Comment (alexkar598) [inline code/__DEFINES/components.dm]: Not used

File: code/modules/mob/emote.dm [hunk 7]
```diff
@@ -13,19 +13,27 @@
...
+		return FALSE
+	var/silenced = FALSE
 	for(var/datum/emote/P in key_emotes)
+		if(!P.check_cooldown(src, intentional))
+			silenced = TRUE
+			continue
 		if(P.run_emote(src, param, m_type, intentional))
-			return
-	if(intentional)
+			SEND_SIGNAL(src, COMSIG_MOB_EMOTE, P, act, m_type, message, intentional)
```
Comment (alexkar598) [inline code/modules/mob/emote.dm]: Not used

**Review by alexkar598 — CHANGES_REQUESTED** [review]

File: code/modules/mob/emote.dm (hunk 7 above)
Comment (Hopekz) [reply code/modules/mob/emote.dm]: Did I do it?

File: code/modules/mob/emote.dm [hunk 8]
```diff
@@ -13,19 +13,27 @@
...
+		return FALSE
+	var/silenced = FALSE
 	for(var/datum/emote/P in key_emotes)
+		if(!P.check_cooldown(src, intentional))
+			silenced = TRUE
+			continue
 		if(P.run_emote(src, param, m_type, intentional))
-			return
-	if(intentional)
+			SEND_SIGNAL(src, P, act, m_type, message, intentional)
```
Comment (alexkar598) [inline code/modules/mob/emote.dm]: No like just delete this line
```suggestion
```

**Pull request merged**